script/tool/lib/src/test_command.dart

```diff
@@ -32,72 +32,64 @@ class TestCommand extends PluginCommand {
-    await for (final Directory packageDir in getPackages()) {
-      final String packageName =
-          p.relative(packageDir.path, from: packagesDir.path);
-      if (!packageDir.childDirectory('test').existsSync()) {
-        print('SKIPPING $packageName - no test subdirectory');
-        continue;
-      }
+  Future<PackageResult> runForPackage(Directory package) async {
+    if (!package.childDirectory('test').existsSync()) {
+      return PackageResult.skip('No test/ directory.');
+    }
 
-      print('RUNNING $packageName tests...');
+    bool passed;
+    if (isFlutterPackage(package)) {
+      passed = await _runFlutterTests(package);
+    } else {
+      passed = await _runDartTests(package);
+    }
+    return passed ? PackageResult.success() : PackageResult.fail();
+  }
 
-      final String enableExperiment = getStringArg(kEnableExperiment);
+  /// Runs the Dart tests for a Flutter package, returning true on success.
+  Future<bool> _runFlutterTests(Directory package) async {
+    final String experiment = getStringArg(kEnableExperiment);
 
-      // `flutter test` automatically gets packages.  `pub run test` does not.  :(
-      int exitCode = 0;
-      if (isFlutterPackage(packageDir)) {
-        final List<String> args = <String>[
-          'test',
-          '--color',
-          if (enableExperiment.isNotEmpty)
-            '--enable-experiment=$enableExperiment',
-        ];
+    final int exitCode = await processRunner.runAndStream(
+      flutterCommand,
+      <String>[
+        'test',
+        '--color',
+        if (experiment.isNotEmpty) '--enable-experiment=$experiment',
+        if (isWebPlugin(package)) '--platform=chrome',
+      ],
+      workingDir: package,
+    );
+    return exitCode == 0;
+  }
 
-        if (isWebPlugin(packageDir)) {
-          args.add('--platform=chrome');
-        }
-        exitCode = await processRunner.runAndStream(
-          'flutter',
-          args,
-          workingDir: packageDir,
-        );
-      } else {
-        exitCode = await processRunner.runAndStream(
-          'dart',
-          <String>['pub', 'get'],
-          workingDir: packageDir,
-        );
-        if (exitCode == 0) {
-          exitCode = await processRunner.runAndStream(
-            'dart',
-            <String>[
-              'pub',
-              'run',
-              if (enableExperiment.isNotEmpty)
-                '--enable-experiment=$enableExperiment',
-              'test',
-            ],
-            workingDir: packageDir,
-          );
-        }
-      }
-      if (exitCode != 0) {
-        failingPackages.add(packageName);
-      }
+  /// Runs the Dart tests for a non-Flutter package, returning true on success.
+  Future<bool> _runDartTests(Directory package) async {
+    // Unlike `flutter test`, `pub run test` does not automatically get
+    // packages
+    int exitCode = await processRunner.runAndStream(
+      'dart',
+      <String>['pub', 'get'],
```

Comment:
(Isn't it slightly odd that flutter tests don't attempt to `flutter pub get` too?)

Reply:
See the comment just above this line.